### ISSUE TITLE
Analog stick interpolation issue with the keyboard

### ISSUE THREAD

**Issue by PiKeyAr** [flyinghead/flycast#334]
Platform / OS / Hardware: x86 / Windows 10

Github hash: a51f310

Hardware: i7 9700K / GTX 1660 Super / 16GB RAM / USB keyboard and mouse

**_Description of the Issue_**

When the analog stick is mapped to keyboard keys, diagonal interpolation often fails to register the correct direction; for example up+right only registers as up. Sometimes it auto-corrects and switches to the right direction after about half a second, but not always. I have recorded a video demonstrating the issue. The video shows flycast booting a controller test utility CDI next to a NohBoard window to visualize what keys are actually being pressed.

**_Additional data_**

Video: https://dcmods.unreliable.network/owncloud/data/PiKeyAr/files/Various/2021-08-29%2015-59-09.mp4
Controller test CDI: https://dcmods.unreliable.network/owncloud/data/PiKeyAr/files/Various/controller_test_2.7z

**Comment by flyinghead** [comment]
I cannot reproduce the issue here using the same controller test CDI. The analog stick responds immediately to any key press, diagonal or not. I haven't tested on Windows though.

Are you running any keyboard utility that could get in the way? Does using RawInput change anything?

**Comment by flyinghead** [comment]
Same on Windows: the analog stick responds to all key presses

**Comment by PiKeyAr** [comment]
Using raw input has no effect on the issue. I'm not running any keyboard utilities besides NohBoard, which is only used to visualize keypresses as the issue happens without it. I can consistently reproduce it by mapping the analog stick to arrow keys, and holding up + left, then adding right and releasing left.
I haven't had this issue on other emulators with the same setup.

**Comment by PiKeyAr** [comment]
Actually just holding keys in the following order reproduces it:
1. Hold Up.
2. Hold Up + Down
3. Release Down but keep holding Up
The cursor will get stuck at the center even though Up is still being held.

**Comment by flyinghead** [comment]
Yes, I can reproduce this last case. However you can't move an analog stick both up and down so it shouldn't be an issue for games.
The analog stick values only change on key up and down events:
L down -> -127
L up -> 0
R down -> 127
R up -> 0
Same for the vertical axis.

So you'll have issues if you hold both up and down or left and right. The correction you see after 500 ms might be the keyboard autorepeat kicking in and sending new key down events.

**Comment by PiKeyAr** [comment]
Thank you for looking into this.


I understand the part about the analog stick being impossible to be held both up and down. However, the current behavior is suboptimal for keyboard players. The last case in particular demonstrates that the input is stuck as if nothing is being pressed even though you're currently holding up, and it won't resolve until you let go of the key, which can confuse the player. I originally noticed this problem in games and thought something was wrong with my keyboard, but this doesn't happen in other emulators or PC games, so I thought I should report it.

**Comment by flyinghead** [comment]
I pushed a fix to this issue on the libretro branch. You might want to give it a try.

**Comment by PiKeyAr** [comment]
Great news! Control handling with the keyboard feels absolutely excellent. Games like Sonic Adventure are much more playable with the keyboard now.
I really appreciate you taking the time to look into the issue and develop a fix so quickly.
